src/cmd/policy.go

```diff
@@ -136,6 +144,61 @@ func RegoFuncGetGithubRepo(ctx rego.BuiltinContext, a, b *ast.Term) (*ast.Term,
 
 	reader := strings.NewReader(response.String())
 	v, err := ast.ValueFromReader(reader)
+
+	return ast.NewTerm(v), nil
+}
+
+func RegoFuncGetGitlabRepo(ctx rego.BuiltinContext, a *ast.Term) (*ast.Term, error) {
+
+	var path string
+	if err := ast.As(a.Value, &path); err != nil {
+		return nil, err
+	}
+
+	if path == "" {
+		log.Error().Msgf("opslevel.repo.gitlab(\"%s\") failed: Please provide a valid Gitlab project path", path)
+		return nil, nil
+	}
+
+	escapedPath := url.QueryEscape(path)
+	gitlabToken := viper.GetString("gitlab-token")
+	gitlabAPIUrl := fmt.Sprintf("https://gitlab.com/api/v4/projects/%s", escapedPath)
+
+	type GitlabRepoMetadata struct {
+		Name        string             `json:"name,omitempty"`
+		Description string             `json:"description,omitempty"`
+		Languages   map[string]float32 `json:"languages,omitempty"`
+	}
+
+	var repoMetadata GitlabRepoMetadata
+
+	response, err := getClientRest().R().
+		SetHeader("PRIVATE-TOKEN", gitlabToken).
+		SetResult(&repoMetadata).
+		Get(gitlabAPIUrl)
+	cobra.CheckErr(err)
+
+	if response.IsError() == true {
+		log.Error().Msgf("error requesting Gitlab repo metadata. CODE: %d: REASON: %s", response.StatusCode(), response)
+		return nil, nil
+	}
+
+	languagesResponse, err := getClientRest().R().
+		SetHeader("PRIVATE-TOKEN", gitlabToken).
+		SetResult(&repoMetadata.Languages).
+		Get(gitlabAPIUrl + "/languages")
+	cobra.CheckErr(err)
+
+	if languagesResponse.IsError() == true {
+		log.Error().Msgf("error requesting Gitlab repo languages. CODE: %d: REASON: %s", response.StatusCode(), response)
+		return nil, nil
+	}
+
+	repoMetadataMarshal, err := json.Marshal(repoMetadata)
+	cobra.CheckErr(err)
+	reader := strings.NewReader(string(repoMetadataMarshal))
+	v, err := ast.ValueFromReader(reader)
```

Comment:
is this the 3rd time we've done this kind of marshaling to AST?  Should be build a reusable function that bundles all this up and takes in an interface (or even better Generics)?

Reply:
Ok